jms/src/main/scala/akka/stream/alpakka/jms/JmsSourceStage.scala

```diff
@@ -5,80 +5,319 @@
+  override def postStop(): Unit = {
+    queue.clear()
+    stopSessions()
+  }
+}
+
+class NoAckLogic(shape: SourceShape[Message],
+                 out: Outlet[Message],
+                 settings: JmsSourceSettings,
+                 attributes: Attributes)
+    extends SourceStageLogic[Message](shape, out, settings, attributes) {
+
+  private val bufferSize = (settings.bufferSize + 1) * settings.sessionCount
+
+  private val backpressure = new Semaphore(bufferSize)
+
+  private[jms] def createSession(connection: Connection, createDestination: Session => javax.jms.Destination) = {
+    val session =
+      connection.createSession(false, settings.acknowledgeMode.getOrElse(AcknowledgeMode.AutoAcknowledge).mode)
+    new JmsSession(connection, session, createDestination(session))
+  }
+
+  private[jms] def pushMessage(msg: Message): Unit = {
+    push(out, msg)
+    backpressure.release()
+  }
+
+  override private[jms] def onSessionOpened(jmsSession: JmsSession): Unit =
+    jmsSession
+      .createConsumer(settings.selector)
+      .onComplete {
+        case Success(consumer) =>
+          consumer.setMessageListener(new MessageListener {
+            def onMessage(message: Message): Unit = {
+              backpressure.acquire()
+              handleMessage.invoke(message)
+            }
+          })
+        case Failure(e) =>
+          fail.invoke(e)
       }
+}
+
+class ExplicitAckLogic(shape: SourceShape[AckEnvelope],
+                       out: Outlet[AckEnvelope],
+                       settings: JmsSourceSettings,
+                       attributes: Attributes)
+    extends SourceStageLogic[AckEnvelope](shape, out, settings, attributes) {
 
-      override private[jms] def onSessionOpened(): Unit =
-        jmsSession.createConsumer(settings.selector).onComplete {
+  private val maxPendingAck = settings.bufferSize
+
+  private[jms] def createSession(connection: Connection, createDestination: Session => javax.jms.Destination) = {
+    val session =
+      connection.createSession(false, settings.acknowledgeMode.getOrElse(AcknowledgeMode.ClientAcknowledge).mode)
+    new JmsAckSession(connection, session, createDestination(session), settings.bufferSize)
+  }
+
+  private[jms] def pushMessage(msg: AckEnvelope): Unit = push(out, msg)
+
+  override private[jms] def onSessionOpened(jmsSession: JmsSession): Unit =
+    jmsSession match {
+      case session: JmsAckSession =>
+        session.createConsumer(settings.selector).onComplete {
           case Success(consumer) =>
             consumer.setMessageListener(new MessageListener {
-              override def onMessage(message: Message): Unit = {
-                backpressure.acquire()
-                handleMessage.invoke(message)
+
+              var listenerStopped = false
+
+              def onMessage(message: Message): Unit = {
+
+                @tailrec
+                def ackQueued(): Unit =
+                  Option(session.ackQueue.poll()) match {
```

Comment:
Minor performance thing, if you can use `akka.util.OptionVal` or just a null check here it saves us an allocation (and subsequent gc) per queued ack.

Reply:
Nice! Let me try that. Thanks for the advice.